WC Tracker: Add info about features and plugin compatibility

coreymckrill opened the pull request:
### Changes proposed in this Pull Request:

* Adds a data key with the list of WC features that are currently enabled on the site.
* Adds a `feature_compat` key to each plugin that shows its compatibility with each feature (compatible, incompatible, or uncertain) if the plugin is "WooCommerce aware", otherwise just an empty array.

### How to test the changes in this Pull Request:

1. Add this snippet to a new file in your test site's `wp-content/mu-plugins` directory:

```php
<?php

add_action(
	'admin_notices',
	function() {
		echo '<div class="notice notice-info"><pre>';
		print_r( \WC_Tracker::get_tracking_data() );
		echo '</pre></div>';
	}
);
```

2. Then view any admin page that will display admin notices (i.e. not block editor screens). You should see an admin notice at the top that outputs all the data collected by WC Tracker.
3. Check for a key called `enabled_features` and see what's listed there. By default it will probably show `analytics` and if you have HPOS enabled, it will also list `custom_order_tables`.
4. Look through the lists of active and inactive plugins. Each should now have an additional `feature_compat` property that lists that plugin's compatibility status with each WooCommerce feature, or, if the plugin is unrelated to WooCommerce, it should just be an empty array.

coreymckrill committed:
WC Tracker: Add enabled features, plugin feature compatibility

* Adds a data key with the list of WC features that are currently
  enabled on the site.
* Adds a `feature_compat` key to each plugin that shows its
  compatibility with each feature (compatibile, incompatible, uncertain)
  if the plugin is "woocommerce aware", otherwise just an empty array.

coreymckrill committed:
Add changelog file

coreymckrill committed:
phpcs cleanup

menakas commented on plugins/woocommerce/includes/class-wc-tracker.php:
```diff
@@ -300,6 +303,10 @@ private static function get_all_plugins() {
 			if ( isset( $v['PluginURI'] ) ) {
 				$formatted['plugin_uri'] = strip_tags( $v['PluginURI'] );
 			}
+			$formatted['feature_compat'] = array();
```
It might be nice to have `feature_compatibility` or `compatible_features` instead of the truncated version.

coreymckrill replied:
👍 Changed in e190bb272e

menakas reviewed (dismissed):
@coreymckrill 
LGTM 👍

coreymckrill committed:
Rename feature_compat to feature_compatibility

vedanshujain reviewed (commented):
Neat work! Just a small suggestion, can you rebase this PR off #38293. This way you will be able to add a unit test for the new data.

deltaWhiskey reviewed (approved):
Looks good!

coreymckrill commented:
@vedanshujain can you clarify what kind of test you're thinking about for this? Both of the functions used here to retrieve and add tracker data, `FeaturesController::get_features` and `FeaturesController::get_compatible_features_for_plugin`, already have [tests](https://github.com/woocommerce/woocommerce/blob/8098c35588b71fd3a8b6a82e7690b96fc0d85a82/plugins/woocommerce/tests/php/src/Internal/Features/FeaturesControllerTest.php) to confirm that they produce the right data.

vedanshujain commented:
@coreymckrill , so I would like to capture the new tracker data somewhere in unit tests, basically we missed adding HPOS compat to rest of tracker data as it wasn't tested anywhere and so nothing in our CI failed. I think an assertion for `feature_compatibility` key in the `data['active_plugins']` would act as a nice integration test, wdyt?

coreymckrill committed:
Add unit tests

coreymckrill committed:
phpcs cleanup

coreymckrill commented:
Looks like the unit test I added is breaking some other unit tests in PluginUtilTests. I'm looking into it...

coreymckrill committed:
Ensure test mocks don't pollute other tests

coreymckrill committed:
Tweak unit tests

coreymckrill committed:
Remove feature compatibility unit test

The mocks in this test were affecting other unit tests for the
PluginUtil class, but only when run in GitHub actions (the tests were
working fine locally). Rather than hold up this PR, I've created
issue #38720 as a future task to add the unit test back in once we've
figured out if there's something buggy happening with dependency
injection.

coreymckrill committed:
Test removal cleanup

coreymckrill commented:
Removed the test and opened an issue to add it later, once dependency injection issues are sorted.
https://github.com/woocommerce/woocommerce/issues/38720
